rfcs/20210119-determinism.md

```diff
@@ -0,0 +1,126 @@
+# RFC: Enabling Determinism in TensorFlow
+  
+| Status        | Proposed                                                                     |
+:---------------|:-----------------------------------------------------------------------------|
+| **Author(s)** | Pankaj Kanwar (Google), Duncan Riach (NVIDIA), Reed Wanderman-Milne (Google) |
+| **Sponsor**   | Sanjoy Das (Google)                                                          |
+| **Updated**   | 2021-03-10                                                                   |
+
+## Objective
+
+Add an API which will make all ops deterministic. This means if a user runs an op multiple times with the same inputs, the outputs will be the same each time. Users can use this API to create deterministic models.
+
+## Motivation
+
+There are several mission critical applications in medicine, finance and automation that require deterministic behavior. Determinism is required so that the behavior of these applications can be accurately predicted & demonstrated in a variety of scenarios.
+
+The lack of determinism in certain ops prevents companies from launching products using models developed in TF. For a subset of these industries having deterministic behavior is a regulatory requirement.
+
+In addition, deterministic ops increases model velocity development by reducing noise, while also simplifying the debugging workflow.
+
+## Design Proposal
+
+We will create a new flag with the default value of "False" which enables deterministic ops.  We will define 2 functions:
+
+* `tf.config.enable_deterministic_ops(enabled)`
+* `tf.config.deterministic_ops_enabled()`
+
+The first function takes in a boolean value, and allows the model developer to enable/disable deterministic ops. The second function returns a bool indicating whether deterministic ops is enabled.
+
+Once enabled, every built-in op will either be made deterministic or raise an error if determinism is not supported. For ops which we have not yet implemented a deterministic version, a `NotImplementedError` will be raised. In the long term, we plan on adding a deterministic version to all such ops. For ops which are inherently nondeterministic such as `tf.random.normal` without a seed, a `FailedPreconditionError` will be raised (the precondition being that determinism must be disabled). Certain ops will only raise an error for certain input shapes or attributes. Depending on the op, in graph mode, the error will either be raised when the op is constructed or when the op is run.
```

Comment:
"For ops which we have not yet implemented a deterministic version, a `NotImplementedError` will be raised" (grammar) -> "A `tf.errors.UnimplementedError` will be raised by ops for which we have not yet implemented a deterministic version."

Also, throughout, `NotImplementedError` -> `tf.errors.UnimplementedError`

"Certain ops will only raise an error for certain input shapes or attributes." -> "Some ops will only raise an error on a subset of input shapes, attributes, data types, or data-paths through the op."